src/passes/DeadArgumentElimination.cpp

```diff
@@ -318,38 +316,48 @@ struct DAE : public Pass {
       if (func->imported()) {
         continue;
       }
-      // We can only optimize if we see all the calls and can modify them.
-      if (hasUnseenCalls[index]) {
+      // References prevent optimization.
+      if (hasRef[index]) {
         continue;
       }
       auto& calls = allCalls[index];
-      if (calls.empty()) {
+      if (calls.empty() && !isExported[index]) {
         // Nothing calls this, so it is not worth optimizing.
         continue;
       }
-      // Refine argument types before doing anything else. This does not
-      // affect whether an argument is used or not, it just refines the type
-      // where possible.
       auto name = func->name;
-      if (refineArgumentTypes(func, calls, module, infoMap[name])) {
-        worthOptimizing.insert(func);
-        markStale(func->name);
+      // Exports prevent refining of argument types, as we don't see all the
+      // calls.
```

Comment:
And also refining the exported function type would allow generalizing the parameter types but not refining them.

Reply:
Yes (but this pass never generalizes).